torch/tensor.py

```diff
@@ -300,7 +302,24 @@ def __matmul__(self, other):
             return self.unsqueeze(0).mm(other).squeeze(0)
         elif dim_self == 2 and dim_other == 2:
             return self.mm(other)
-        raise ValueError("both arguments to __matmul__ need to be 1D or 2D, "
+        elif dim_self >= 2 and dim_other >= 2:
+            # ensure each tensor is at least 3-dimensional
+            self_exp_size = torch.Size((1,) * max(3 - self.dim(), 0) + self.size())
+            other_exp_size = torch.Size((1,) * max(3 - other.dim(), 0) + other.size())
+
+            # expand the batch portion (i.e. cut off matrix dimensions and expand rest)
+            expand_batch_portion = torch._C._infer_size(self_exp_size[:-2], other_exp_size[:-2])
+
+            # flatten expanded batches
+            self_expanded = self.expand(*(expand_batch_portion + self_exp_size[-2:])) \
+                .contiguous().view(reduce(mul, expand_batch_portion), *self_exp_size[-2:])
+            other_expanded = other.expand(*(expand_batch_portion + other_exp_size[-2:])) \
+                .contiguous().view(reduce(mul, expand_batch_portion), *other_exp_size[-2:])
+
+            # reshape batches back into result
+            total_expansion = expand_batch_portion + (self_exp_size[-2], other_exp_size[-1])
+            return self_expanded.bmm(other_expanded).view(*(total_expansion))
```

Comment:
Agreed, this is a good idea.  I'm going to open a separate issue for it given this issue is large enough and ready to merge already.